From 4027c24a3b58461b5552bb3002b769234abca0d9 Mon Sep 17 00:00:00 2001
From: Karthikeyan Kathirvel <quic_kathirve@quicinc.com>
Date: Thu, 7 Apr 2022 12:21:44 +0530
Subject: [PATCH] ath11k: skip caldata download with module param

Skip the caldata download to ignore loading corrupted caldata error

This patch would help probe interfaces without FW assert though the
caldata is corrupted. It is useful for initial calibration.
Interfaces cannot be brought up since the device should be in
FTM mode.

insmod ath11k.ko ftm_mode=1 skip_caldata=1

Downloading of caldata can be skipped from u-boot as well

setenv ath11k_mod "skip_caldata=1 ftm_mode=1"

Signed-off-by: Karthikeyan Kathirvel <quic_kathirve@quicinc.com>
---
 drivers/net/wireless/ath/ath11k/qmi.c | 9 +++++++++
 1 file changed, 9 insertions(+)

--- a/drivers/net/wireless/ath/ath11k/qmi.c
+++ b/drivers/net/wireless/ath/ath11k/qmi.c
@@ -212,6 +212,10 @@ static struct qmi_elem_info qmi_wlanfw_q
 	},
 };
 
+static bool ath11k_skip_caldata;
+module_param_named(skip_caldata, ath11k_skip_caldata, bool, 0444);
+MODULE_PARM_DESC(ath11k_skip_caldata, "Skip caldata download");
+
 static struct qmi_elem_info qmi_wlanfw_host_cap_req_msg_v01_ei[] = {
 	{
 		.data_type	= QMI_OPT_FLAG,
@@ -3279,6 +3283,16 @@ static int ath11k_qmi_load_bdf_qmi(struc
 	if (bdf_type == ATH11K_QMI_BDF_TYPE_ELF)
 		goto out;
 
+	if (ath11k_skip_caldata) {
+		if (ath11k_ftm_mode) {
+			ath11k_warn(ab, "Skipping caldata download in FTM mode\n");
+			goto out;
+		}
+		ath11k_err(ab, "failed to skip caldata download. FTM mode is not enabled\n");
+		ret = -EOPNOTSUPP;
+		goto out;
+	}
+
 	if (ab->qmi.target.eeprom_caldata) {
 		file_type = ATH11K_QMI_FILE_TYPE_EEPROM;
 		tmp = filename;
